File: ts/sb/struct_recharge_a.ts
import * as _ from "./_"
import * as TplEnum from "./enum"

export interface RechargeA extends _.Serializable, _.Deserializable {
    id: number;
    type: TplEnum.OrderStatus[];
    phone: string[];
    si: _.SimInfo | null;
    aid: number;
}

export const newRechargeA = (): RechargeA => {
    const s = {
        id: 0,
        type: [],
        phone: [],
        si: null,
        aid: 0,
    } as any as RechargeA;
    s.set = (buf: _.Buffer) => setRechargeA(buf, s);
    s.get = (buf: _.Buffer) => {
        const [res, err] = getRechargeA(buf);
        if (err === null) Object.assign(s, res);
        return err;
    };
    return s;
}

export const eqRechargeA = (a: RechargeA, b: RechargeA): boolean => {
    if (a === b) return true;
    if (a === null || b === null) return false;
    if (!_.eqU32(a.id, b.id)) return false;
    if (!_.eqU8List(a.type as any, b.type as any)) return false;
    if (!_.eqTextList(a.phone, b.phone)) return false;
    if ((a.si === null) !== (b.si === null)) return false;
    if (a.si !== null && b.si !== null && !_.eqSimInfo(a.si, b.si)) return false;
    if (!_.eqU32(a.aid, b.aid)) return false;
    return true;
}

export const getRechargeA = (buf: _.Buffer): [RechargeA, Error | null] => {
    const s = newRechargeA();
    const bitmaskSize = 1;
    const [bits, err] = buf.read(bitmaskSize);
    if (err !== null) return [s, err];
    if (_.GetBit(bits, 0)) {
        const [v, err] = _.getU32(buf);
        if (err !== null) return [s, err];
        s.id = v;
    }
    if (_.GetBit(bits, 1)) {
        const [v, err] = _.getU8List(buf);
        if (err !== null) return [s, err];
        s.type = v as any;
        if (!_.IsOrderStatusList(s.type as any)) return [s, new Error("get RechargeA Type: invalid enum value")];
    }
    if (_.GetBit(bits, 2)) {
        const [v, err] = _.getTextList(buf);
        if (err !== null) return [s, err];
        s.phone = v;
    }
    if (_.GetBit(bits, 3)) {
        const [v, err] = _.getSimInfo(buf);
        if (err !== null) return [s, err];
        s.si = v;
    }
    if (_.GetBit(bits, 4)) {
        const [v, err] = _.getU32(buf);
        if (err !== null) return [s, err];
        s.aid = v;
    }
    return [s, null];
}

export const setRechargeA = (buf: _.Buffer, s: RechargeA): Error | null => {
    if (s === null || s === undefined) return new Error(`set RechargeA: value is null or undefined`);
    const startOffset = buf.write_offset;
    const bits = new Uint8Array(1);
    if (!_.eqU32(s.id, 0)) {
        _.SetBit(bits, 0, true);
    }
    if (s.type && s.type.length > 0) {
        if (!_.IsOrderStatusList(s.type as any)) return new Error("set RechargeA Type: invalid enum value");
        _.SetBit(bits, 1, true);
    }
    if (s.phone && s.phone.length > 0) {
        _.SetBit(bits, 2, true);
    }
    if (s.si !== null && s.si !== undefined) {
        _.SetBit(bits, 3, true);
    }
    if (!_.eqU32(s.aid, 0)) {
        _.SetBit(bits, 4, true);
    }

    const errBits = buf.write(bits);
    if (errBits !== null) return errBits;
    if (!_.eqU32(s.id, 0)) {
        const err = _.setU32(buf, s.id);
        if (err !== null) {
            buf.rewindWrite(startOffset);
            return err;
        }
    }
    if (s.type && s.type.length > 0) {
        const err = _.setU8List(buf, s.type as any);
        if (err !== null) {
            buf.rewindWrite(startOffset);
            return err;
        }
    }
    if (s.phone && s.phone.length > 0) {
        const err = _.setTextList(buf, s.phone);
        if (err !== null) {
            buf.rewindWrite(startOffset);
            return err;
        }
    }
    if (s.si !== null && s.si !== undefined) {
        const err = _.setSimInfo(buf, s.si);
        if (err !== null) {
            buf.rewindWrite(startOffset);
            return err;
        }
    }
    if (!_.eqU32(s.aid, 0)) {
        const err = _.setU32(buf, s.aid);
        if (err !== null) {
            buf.rewindWrite(startOffset);
            return err;
        }
    }
    return null;
}

export const getRechargeAList = (buf: _.Buffer): [RechargeA[], Error | null] => {
    const [count, err] = _.getU16(buf);
    if (err !== null) return [[], err];
    const list: RechargeA[] = new Array(count);
    for (let i = 0; i < count; i++) {
        const [item, err2] = getRechargeA(buf);
        if (err2 !== null) return [[], err2];
        list[i] = item;
    }
    return [list, null];
}
export const setRechargeAList = (buf: _.Buffer, v: RechargeA[]): Error | null => {
    if (v.length > 65535) return new Error(`list length ${v.length} exceeds u16 max`);
    const err = _.setU16(buf, v.length);
    if (err !== null) return err;
    for (const item of v) {
        const err2 = setRechargeA(buf, item);
        if (err2 !== null) return err2;
    }
    return null;
}
export const eqRechargeAList = (a: RechargeA[], b: RechargeA[]): boolean => _.eqList(a, b, eqRechargeA);

File: go/sb/struct_recharge_a.go
package sb

import (
	"bytes"
	"fmt"
	"slices"
)

type RechargeA struct {
	// abcd
	Id uint32 `bson:"_id" json:"_id"`
	Type []OrderStatus `bson:"type" json:"type"`
	Phone []string `bson:"phone" json:"phone"`
	Si *SimInfo `bson:"si" json:"si"`
	Aid uint32 `bson:"aid" json:"aid"`
}

func sizeRechargeABody(s *RechargeA, bits []byte) (int, error) {
	if s == nil { return 0, fmt.Errorf("sizeRechargeA: nil value") }
	bodySize := 0
	if s.Id != 0 {
		bodySize += 4
		if bits != nil { SetBit(bits, uint8(0), true) }
	}
	if len(s.Type) > 0 {
		fieldSize, err := sizeFixedList(len(s.Type), 1)
		if err != nil { return 0, fmt.Errorf("sizeRechargeA Type: %w", err) }
		bodySize += fieldSize
		if bits != nil { SetBit(bits, uint8(1), true) }
	}
	if len(s.Phone) > 0 {
		fieldSize, err := sizeTextList(s.Phone)
		if err != nil { return 0, fmt.Errorf("sizeRechargeA Phone: %w", err) }
		bodySize += fieldSize
		if bits != nil { SetBit(bits, uint8(2), true) }
	}
	if s.Si != nil {
		fieldSize, err := sizeSimInfo(s.Si)
		if err != nil { return 0, fmt.Errorf("sizeRechargeA Si: %w", err) }
		bodySize += fieldSize
		if bits != nil { SetBit(bits, uint8(3), true) }
	}
	if s.Aid != 0 {
		bodySize += 4
		if bits != nil { SetBit(bits, uint8(4), true) }
	}
	return bodySize, nil
}

func sizeRechargeA(s *RechargeA) (int, error) {
	bodySize, err := sizeRechargeABody(s, nil)
	if err != nil { return 0, err }
	return 1 + bodySize, nil
}

func sizeRechargeAList(v RechargeAList) (int, error) {
	if len(v) > 65535 { return 0, fmt.Errorf("list length exceeds uint16 max") }
	totalSize := 2
	for i, item := range v {
		itemSize, err := sizeRechargeA(item)
		if err != nil { return 0, fmt.Errorf("sizeRechargeAList[%d]: %w", i, err) }
		totalSize += itemSize
	}
	return totalSize, nil
}

func GetRechargeA(buf *bytes.Buffer, s *RechargeA) error {
	if s == nil { return nil }
	const bitSize = 1
	if buf.Len() < bitSize { return fmt.Errorf("GetRechargeA bitmask: %d - %d", buf.Len(), bitSize) }
	bits := buf.Next(bitSize)
	if GetBit(bits, uint8(0)) {
		val, err := GetU32(buf)
		if err != nil { return fmt.Errorf("GetRechargeA Id: %w", err) }
		s.Id = val
	}
	if GetBit(bits, uint8(1)) {
		val, err := GetOrderStatusList(buf)
		if err != nil { return fmt.Errorf("GetRechargeA Type: %w", err) }
		s.Type = val
	}
	if GetBit(bits, uint8(2)) {
		val, err := GetTextList(buf)
		if err != nil { return fmt.Errorf("GetRechargeA Phone: %w", err) }
		s.Phone = val
	}
	if GetBit(bits, uint8(3)) {
		val, err := ReadSimInfo(buf)
		if err != nil { return fmt.Errorf("GetRechargeA Si: %w", err) }
		s.Si = val
	}
	if GetBit(bits, uint8(4)) {
		val, err := GetU32(buf)
		if err != nil { return fmt.Errorf("GetRechargeA Aid: %w", err) }
		s.Aid = val
	}
	if err := ValidateRechargeA(s); err != nil { return fmt.Errorf("ValidateRechargeA: %w", err) }
	return nil
}

func SetRechargeA(buf *bytes.Buffer, s *RechargeA) error {
	if s == nil { return fmt.Errorf("SetRechargeA: nil value") }
	if err := ValidateRechargeA(s); err != nil { return fmt.Errorf("ValidateRechargeA: %w", err) }
	const bitSize = 1
	var bits [1]byte
	bodySize, err := sizeRechargeABody(s, bits[:])
	if err != nil { return err }
	buf.Grow(bitSize + bodySize)
	if _, err := buf.Write(bits[:]); err != nil { return fmt.Errorf("SetRechargeA write bitmask: %w", err) }
	if s.Id != 0 {
		if err := SetU32(buf, s.Id); err != nil { return fmt.Errorf("SetRechargeA Id: %w", err) }
	}
	if len(s.Type) > 0 {
		if err := SetOrderStatusList(buf, (OrderStatusList)(s.Type)); err != nil { return fmt.Errorf("SetRechargeA Type: %w", err) }
	}
	if len(s.Phone) > 0 {
		if err := SetTextList(buf, s.Phone); err != nil { return fmt.Errorf("SetRechargeA Phone: %w", err) }
	}
	if s.Si != nil {
		if err := SetSimInfo(buf, s.Si); err != nil { return fmt.Errorf("SetRechargeA Si: %w", err) }
	}
	if s.Aid != 0 {
		if err := SetU32(buf, s.Aid); err != nil { return fmt.Errorf("SetRechargeA Aid: %w", err) }
	}
	return nil
}

func ValidateRechargeA(s *RechargeA) error {
	if s == nil { return nil }
	for i, item := range s.Type {
		if !IsOrderStatus(item) { return fmt.Errorf("Type[%d] 非法枚举值: %d", i, item) }
	}
	if s.Si != nil {
		if err := ValidateSimInfo(s.Si); err != nil { return fmt.Errorf("Si: %w", err) }
	}
	return nil
}

func EqRechargeA(a, b *RechargeA) bool {
	if a == b { return true }
	if a == nil || b == nil { return false }
	if !EqU32(a.Id, b.Id) { return false }
	if !EqOrderStatusList(a.Type, b.Type) { return false }
	if !EqTextList(a.Phone, b.Phone) { return false }
	if !EqSimInfo(a.Si, b.Si) { return false }
	if !EqU32(a.Aid, b.Aid) { return false }
	return true
}

// Standalone functions
func ReadRechargeA(buf *bytes.Buffer) (*RechargeA, error) {
	s := new(RechargeA)
	return s, GetRechargeA(buf, s)
}

type RechargeAList []*RechargeA
func GetRechargeAList(buf *bytes.Buffer) (RechargeAList, error) {
	count, err := GetU16(buf)
	if err != nil { return nil, err }
	list := make([]*RechargeA, count)
	for i := range list {
		item, err := ReadRechargeA(buf)
		if err != nil { return nil, err }
		list[i] = item
	}
	return RechargeAList(list), nil
}
func SetRechargeAList(buf *bytes.Buffer, v RechargeAList) error {
	totalSize, err := sizeRechargeAList(v)
	if err != nil { return err }
	buf.Grow(totalSize)
	if err := SetU16(buf, uint16(len(v))); err != nil { return err }
	for _, item := range v {
		if err := SetRechargeA(buf, item); err != nil { return err }
	}
	return nil
}
func ValidateRechargeAList(v RechargeAList) error {
	for i, item := range v {
		if item == nil { return fmt.Errorf("RechargeAList[%d]: nil item", i) }
		if err := ValidateRechargeA(item); err != nil { return fmt.Errorf("RechargeAList[%d]: %w", i, err) }
	}
	return nil
}
func EqRechargeAList(a, b RechargeAList) bool { return slices.EqualFunc(a, b, EqRechargeA) }
